File: webhook/webhook.py
from flask import Flask, request, jsonify
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.padding import PKCS7

from base64 import b64decode
import hashlib
import json

app = Flask(__name__)

# Configuration for AES
AES_KEY = b'oFcTbSzdmOzXF2DZ72h9ILdLsXTE8qK1'
IV = b'IVa7b9f8d6e54gha'

def decrypt_payload(encrypted_data):
    backend = default_backend()
    cipher = Cipher(algorithms.AES(AES_KEY), modes.CBC(IV), backend=backend)
    decryptor = cipher.decryptor()
    decrypted_padded_data = decryptor.update(b64decode(encrypted_data)) + decryptor.finalize()
    
    # Handle PKCS7 padding
    unpadder = PKCS7(algorithms.AES.block_size).unpadder()
    decrypted_data = unpadder.update(decrypted_padded_data) + unpadder.finalize()
    return decrypted_data.decode('utf-8')

def verify_signature(payload, signature, timestamp, project_id):
    message = f"{project_id}{AES_KEY.decode()}{timestamp}{payload}"
    calculated_hash = hashlib.sha256(message.encode()).hexdigest()
    return calculated_hash == signature

@app.route('/webhook', methods=['POST'])
def webhook():
    data = request.get_json()
    try:
        print('Received webhook:', data)
        if not verify_signature(data['payload'], data['signature'], data['timestamp'], data['project_id']):
            print('Invalid signature')
            return jsonify({"error": "Invalid signature"}), 403
        
        decrypted_message = decrypt_payload(data['payload'])
        print('Decrypted payload:', decrypted_message)
        
        return jsonify({"message": "ok"}), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

if __name__ == '__main__':
    app.run(port=3000)

File: webhook/webhook.js
const express = require('express');
const bodyParser = require('body-parser');
const crypto = require('crypto');

const app = express();
const PORT = 3000;

// Replace 'aesKey' and 'iv' with your actual AES key and IV
const aesKey = Buffer.from('oFcTbSzdmOzXF2DZ72h9ILdLsXTE8qK1', 'utf-8');  // Ensure your key length matches AES-256 if using that
const iv = Buffer.from('IVa7b9f8d6e54gha', 'utf-8');  // Ensure this matches the IV length (16 bytes for AES)

// Middleware to parse JSON bodies
app.use(bodyParser.json());

function decryptPayload(encryptedData) {
  // Decode from Base64 to get the encrypted bytes
  const encryptedBytes = Buffer.from(encryptedData, 'base64');
  
  // Create the decipher with the AES key and IV
  let decipher = crypto.createDecipheriv('aes-256-cbc', aesKey, iv);
  
  // Decrypt the data
  let decrypted = decipher.update(encryptedBytes, null, 'utf-8');
  decrypted += decipher.final('utf-8');
  return decrypted;
}

function verifySignature(payload, signature, timestamp, projectId) {
  const data = `${projectId}${aesKey}${timestamp}${payload}`;
  const hash = crypto.createHash('sha256').update(data).digest('hex');
  
  return hash === signature;
}


// Handle POST requests to the /webhook endpoint
app.post('/webhook', (req, res) => {
    console.log('Received webhook:', req.body);
    
    // Verify the signature
    if (!verifySignature(req.body.payload, req.body.signature, req.body.timestamp, req.body.project_id)) {
      console.log('Invalid signature');
      return res.status(403).send('Invalid signature');
    }

    // Assuming encrypted data comes as a hex string in the payload field
    const decryptedMessage = decryptPayload(req.body.payload);
    console.log('Decrypted payload:', decryptedMessage);



    // Respond with a status of 200 and a simple message
    res.status(200).send('ok');
});

// Start the server
app.listen(PORT, () => {
    console.log(`Server is running on http://localhost:${PORT}`);
});
